### PR TITLE
Benchmark Enhancements and Code Cleanup

### PR DESCRIPTION
**Pull request by gojimmypi**
## Description

Building on the updates in [#5800](https://github.com/wolfSSL/wolfssl/pull/5800), this PR improves the [wolfSSL wolfcrypt benchmark utility](https://github.com/wolfSSL/wolfssl/tree/master/wolfcrypt/benchmark).

This PR mostly addresses features related to the `GENERATE_MACHINE_PARSEABLE_REPORT` option, and adds some fixed units (GB, MB, KB, or bytes) rather than auto-scaled units.

## Fixes zd# 

(n/a)

## Changes

This PR Fixes [#5814](https://github.com/wolfSSL/wolfssl/issues/5814) Compile-time error when defining `GENERATE_MACHINE_PARSEABLE_REPORT`.

The [incremental changes to the `benchmark.c` file](https://github.com/gojimmypi/wolfssl/commits/Espressif_Multichip/wolfcrypt/benchmark/benchmark.c) are mixed in with my other [Espressif muilt-chip branch](https://github.com/wolfSSL/wolfssl/compare/master...gojimmypi:wolfssl:Espressif_Multichip).

Specific changes:

- Add benchmark cycle counts for Espressif Xtensa architecture on ESP32 chips.
- Adds internal `__BENCHMARK_VALUE_TO_STRING(x)` and `__BENCHMARK_VALUE(x)` macros
- Extends the Espressif [xthal_get_ccount()](https://github.com/espressif/esp-idf/blob/f90c12fc3cc5abb47a7f1088ada43a21b31d0357/components/xtensa/include/xtensa/hal.h#L726) to be _single_-rollover tolerant. See `xthal_get_ccount_ex()`
- Allows for fixed-unit specification via the new `#define` options (in this order, if several supplied):
    `WOLFSSL_BENCHMARK_FIXED_UNITS_G` or `WOLFSSL_BENCHMARK_FIXED_UNITS_GB` for GB/GiB units

    `WOLFSSL_BENCHMARK_FIXED_UNITS_M` or `WOLFSSL_BENCHMARK_FIXED_UNITS_MB` for MB/MiB units 

    `WOLFSSL_BENCHMARK_FIXED_UNITS_K` or `WOLFSSL_BENCHMARK_FIXED_UNITS_KB` for KB/KiB units

    `WOLFSSL_BENCHMARK_FIXED_UNITS_B` or `WOLFSSL_BENCHMARK_FIXED_UNITS_B` for Byte units

- Add support for Expressif Xtensa architecture for `GENERATE_MACHINE_PARSEABLE_REPORT` 
- Properly implement `HAVE_GET_CYCLES` for showing cycle counts on platforms that are known to have this capability.
- Change 'GENERATE_MACHINE_PARSEABLE_REPORT` no longer limits to 80 char lines.
- Define a new internal `__BENCHMARK_MAXIMUM_LINE_LENGTH` to define the maximum allowed line length of each metric. 
- The new maximum line length has been increased from 128 to 150. Also fixes missing LF on some large byte metrics when using `GENERATE_MACHINE_PARSEABLE_REPORT`.
- Properly honor the definition (or lack thereof) the `HAVE_GET_CYCLES`.
- [Code formatting cleanup](https://github.com/gojimmypi/wolfssl/commit/5eec12458e1c0e6ed9160bc4788e9def2f7cd9a3) for wolfSSL coding standards.

## Example Usage

This example will set all the resultant units to KB/KiB and generate a machine-parseable report:
```
./configure --enable-all CFLAGS="-DWOLFSSL_BENCHMARK_FIXED_UNITS_KB -DGENERATE_MACHINE_PARSEABLE_REPORT"
```

## Testing and Review of Changes

Some source code formatting-only changes were untestable and will depend on Jenkins.

Visual inspection: the "cycles" count should be approximately equal to the CPU clock frequency, given that most measurements are for about 1 second duration.

This script may be helpful in testing various changes and parameters of the `benchmark` utility:

```
#!/bin/bash

# sample init:
# git clone https://github.com/wolfSSL/wolfssl.git wolfssl_test
# cd wolfssl_test
# ./autogen.sh

function check_for_error
{
retVal=$1
if [ $retVal -ne 0 ]; then
    echo "Error"
    exit $retVal
fi
}

function test_benchmark # param #1 is CFLAGS value, parameter #2 is output file
{
    echo ./configure --enable-all --enable-jobserver CFLAGS="$1"
         ./configure --enable-all --enable-jobserver CFLAGS="$1"
    check_for_error $?

    echo "Running make..."
    make
    check_for_error $?

    echo Running benchmark...
    ./wolfcrypt/benchmark/benchmark | tee "$2"
    check_for_error $?
}



test_benchmark "-DWOLFSSL_BENCHMARK_FIXED_UNITS_GB" "benchmark_gb_test.log"

test_benchmark "-DWOLFSSL_BENCHMARK_FIXED_UNITS_MB" "benchmark_mb_test.log"

test_benchmark "-DWOLFSSL_BENCHMARK_FIXED_UNITS_KB" "benchmark_kb_test.log"

test_benchmark "-DWOLFSSL_BENCHMARK_FIXED_UNITS_BB" "benchmark_bb_test.log"


test_benchmark "-DWOLFSSL_BENCHMARK_FIXED_UNITS_GB -DGENERATE_MACHINE_PARSEABLE_REPORT" "benchmark_parse_gb_test.log"

test_benchmark "-DWOLFSSL_BENCHMARK_FIXED_UNITS_MB -DGENERATE_MACHINE_PARSEABLE_REPORT" "benchmark_parse_mb_test.log"

test_benchmark "-DWOLFSSL_BENCHMARK_FIXED_UNITS_KB -DGENERATE_MACHINE_PARSEABLE_REPORT" "benchmark_parse_kb_test.log"

test_benchmark "-DWOLFSSL_BENCHMARK_FIXED_UNITS_BB -DGENERATE_MACHINE_PARSEABLE_REPORT" "benchmark_parse_bb_test.log"

test_benchmark "-DWOLFSSL_BENCHMARK_FIXED_UNITS_GB -DGENERATE_MACHINE_PARSEABLE_REPORT -DWOLFSSL_BENCHMARK_FIXED_CSV" "benchmark_parse_csv_gb_test.log"

test_benchmark "-DWOLFSSL_BENCHMARK_FIXED_UNITS_MB -DGENERATE_MACHINE_PARSEABLE_REPORT -DWOLFSSL_BENCHMARK_FIXED_CSV" "benchmark_parse_csv_mb_test.log"

test_benchmark "-DWOLFSSL_BENCHMARK_FIXED_UNITS_KB -DGENERATE_MACHINE_PARSEABLE_REPORT -DWOLFSSL_BENCHMARK_FIXED_CSV" "benchmark_parse_csv_kb_test.log"

test_benchmark "-DWOLFSSL_BENCHMARK_FIXED_UNITS_BB -DGENERATE_MACHINE_PARSEABLE_REPORT -DWOLFSSL_BENCHMARK_FIXED_CSV" "benchmark_parse_csv_bb_test.log"

```

## Not addressed

- There are Intel and Xtensa counts, other platforms may be missing? See `BEGIN/END_INTEL_CYCLES` and `BEGIN/END_ESP_CYCLES` macros.

- The `printf` statements should eventually all be changed to lower-memory-utilization outputs for embedded devices, such as the Espressif `ESP_LOGI().`

- The output examples of turning on `GENERATE_MACHINE_PARSEABLE_REPORT` or not: these formats are similar and could possibly be improved:

```text
(off) RNG                        110 MiB took 1.007 seconds,  109.276 MiB/s Cycles per byte =  26.18
 (on) RNG                        100 MiB took 1.017 seconds,   98.286 MiB/s, 3052336385 cycles, Cycles per byte =  29.11
```

- It might be useful to include other details on the `GENERATE_MACHINE_PARSEABLE_REPORT` lines such as CPU/Platform type, 
date of test, and more... with the intention of putting all this information in a common database.

## Known issues

The units of GB are supported, but any resulting fractional numbers less than 1 are shown as zero.

New platforms may have a zero displayed for CPU cycle counts.

## Developers Tools:

Developers of ESP32 testing for other platforms may find this helpful:

As the application version of the `benchmark.c` file is in the `components` directory of the ESP-IDF,
it needs to be copied back to the local repo when testing host-based benchmark. In this case on WSL/Ubuntu on Windows 10
with the VisualGDB extension to Visual Studio:

```bash
#!/bin/bash

cp /mnt/c/SysGCC/esp32/esp-idf/v4.4.2/components/wolfssl/wolfcrypt/benchmark/benchmark.c wolfcrypt/benchmark/benchmark.c
make

 ./wolfcrypt/benchmark/benchmark

retVal=$1
if [ $retVal -ne 0 ]; then
    echo "Error"
    exit $retVal
fi
```

## Samples

[Benchmark_Espressif_ESP32_sample.log](https://github.com/wolfSSL/wolfssl/files/10189777/Benchmark_Espressif_ESP32_sample.log)

[benchmark_bb_test.log](https://github.com/wolfSSL/wolfssl/files/10189758/benchmark_bb_test.log)
[benchmark_kb_test.log](https://github.com/wolfSSL/wolfssl/files/10189756/benchmark_kb_test.log)
[benchmark_mb_test.log](https://github.com/wolfSSL/wolfssl/files/10189757/benchmark_mb_test.log)
[benchmark_gb_test.log](https://github.com/wolfSSL/wolfssl/files/10189755/benchmark_gb_test.log)

[benchmark_parse_bb_test.log](https://github.com/wolfSSL/wolfssl/files/10189750/benchmark_parse_bb_test.log)
[benchmark_parse_kb_test.log](https://github.com/wolfSSL/wolfssl/files/10189752/benchmark_parse_kb_test.log)
[benchmark_parse_mb_test.log](https://github.com/wolfSSL/wolfssl/files/10189753/benchmark_parse_mb_test.log)
[benchmark_parse_gb_test.log](https://github.com/wolfSSL/wolfssl/files/10189773/benchmark_parse_gb_test.log)

## Checklist

 - [ ] added tests
 - [ ] updated/added doxygen
 - [ ] updated appropriate READMEs
 - [ ] Updated manual and documentation

### CODE REVIEW

**Comment by tim-weller-wolfssl** [comment]
The `-DWOLFSSL_BENCHMARK_FIXED_UNITS_BB`  option looks odd to me.  Following from `MB` == "MegaBytes", 'KB' == "KiloBytes", I read `BB` as "BytesBytes" (?!?).  I suggest removing the `FIXED_UNITS_BB` option, as the `FIXED_UNITS_B` options is sufficient to specify "Bytes" as the fixed unit.

**Review by douzzer — DISMISSED** [review]
This looks like a very nice cleanup!  Good stuff.

It would have been better to keep the substantive changes and the whitespace/linelength cleanup in separate commits.  To cut down the noise for the 1st pass review, I had to use
```
git diff --word-diff-regex='[^[:space:]]' master
```
which worked really well, but still it would've been better to keep the changes separate at the commit level.

I ran your patched benchmark.c through `benchmark-wolfcrypt-intelasm-all` and everything worked right, which is a good test because of the slew of asserts around benchmark data format/integrity there.

In summary, the only needed tweak I found is the column header metarecords.

**Review by tim-weller-wolfssl — CHANGES_REQUESTED** [review]

**Comment by tim-weller-wolfssl** [comment]
The clean-up looks nice!  I just had a few suggestions for consideration.  Thanks!

File: wolfcrypt/benchmark/benchmark.c [hunk 1]
```diff
@@ -968,8 +984,88 @@ static const char* bench_desc_words[][15] = {
     #define SHOW_INTEL_CYCLES_CSV(b, n, s) \
         (void)XSNPRINTF(b + XSTRLEN(b), n - XSTRLEN(b), "%.6f,\n", \
             (float)total_cycles / (count*s))
+#elif defined(WOLFSSL_ESPIDF)
+    static THREAD_LS_T unsigned long long begin_cycles;
```
[the remark this answers is not in the document]
Comment (gojimmypi) [reply wolfcrypt/benchmark/benchmark.c:988]: ah yes, I was seeing this error for Espressif sending data to `XSNPRINTF`:
```
Error    format '%lu' expects argument of type 'long unsigned int', but argument 9 has type 'word64' {aka 'long long unsigned int'} 
```
I ended up casting the value as  `(unsigned long)`.  I'll revert those back to `word64`s.

File: wolfcrypt/benchmark/benchmark.c [hunk 2]
```diff
@@ -1562,29 +1665,140 @@ static WC_INLINE void bench_stats_start(int* count, double* start)
...
+    *blocks /= (1024UL * 1024UL);
+    rt = "MiB";
+
+#elif (   defined (WOLFSSL_BENCHMARK_FIXED_UNITS_K) \
+       || defined (WOLFSSL_BENCHMARK_FIXED_UNITS_KB))
+    *blocks /= 1024;
+    rt = "KiB";
+
+#elif (   defined (WOLFSSL_BENCHMARK_FIXED_UNITS_B)  \
+       || defined (WOLFSSL_BENCHMARK_FIXED_UNITS_BB) )
```
[the remark this answers is not in the document]
Comment (gojimmypi) [reply wolfcrypt/benchmark/benchmark.c:1705]: yes, I agree. Removed in upcoming revision. I got a bit too carried away with the pleasure of consistency.

File: wolfcrypt/benchmark/benchmark.c [hunk 3]
```diff
@@ -1828,7 +2113,8 @@ static void bench_stats_pq_asym_finish(const char* algo, int useDeviceID, int co
         /* only print out header once */
         if (pqasym_header_printed == 0) {
 #ifdef GENERATE_MACHINE_PARSEABLE_REPORT
-            printf("%s", "\"pq_asym\",Algorithm,avg ms,ops/sec,ops,secs,cycles,"
+            printf("%s",
+                   "\"pq_asym\",Algorithm,avg ms,ops/sec,ops,secs,cycles,"
```
[the remark this answers is not in the document]
Comment (gojimmypi) [reply wolfcrypt/benchmark/benchmark.c:2133]: Thanks @douzzer for your review and feedback. Ya, I had contemplated pulling some of the cosmetic changes into a separate PR vs all one monolith.  Noted for future reference. 

I also made the change that @tim-weller-wolfssl suggestion, although technically both the single-character and two-character [fixed-units options](https://github.com/gojimmypi/wolfssl/blob/9d98cd6e5092fe83a2c743d5297a627ab2ab2e6c/wolfcrypt/benchmark/benchmark.c#L1689) will work. Actually, I had forgotten to add to the README, too.

I've made some changes to the PR. Let me know what you think.

[the remark this answers is not in the document]
Comment (gojimmypi) [reply wolfcrypt/benchmark/benchmark.c:2154]: @douzer I had forgotten the conditional cycle PQ header. I've addressed [this](https://github.com/gojimmypi/wolfssl/blob/8831fbd22e50d3340150066971e3c2385d83d6b0/wolfcrypt/benchmark/benchmark.c#L2151) and some other fine tuning in [83d6b0](https://github.com/wolfSSL/wolfssl/pull/5871/commits/8831fbd22e50d3340150066971e3c2385d83d6b0)

File: wolfcrypt/benchmark/benchmark.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (gojimmypi) [reply wolfcrypt/benchmark/benchmark.c]: @tim-weller-wolfssl thanks for your help in testing. I've just pushed a [correction](https://github.com/wolfSSL/wolfssl/pull/5871/commits/8831fbd22e50d3340150066971e3c2385d83d6b0) to the PQ algo header and some other minor changes. Perhaps you can test again? If you don't have PQ installed, see [docs](https://github.com/wolfSSL/wolfssl/blob/a3f3c76faa4e5125415df9b80ebca80cf5e53946/INSTALL#L139).

**Comment by gojimmypi** [comment]
I've manually merged this PR with the recent changes from https://github.com/wolfSSL/wolfssl/pull/5646

@tim-weller-wolfssl with these changes I've re-requested a review from you, please.

@JacobBarthelmeh perhaps you could check to confirm I merged your changes correctly? Thank you.

**Review by tim-weller-wolfssl — APPROVED** [review]
I re-ran the test script provided and the results look reasonable!

**Review by dgarske — APPROVED** [review]
Great cleanup! Tested all various build options including the async version. Thanks Jim.
